FABulous/fabric_generator/file_parser.py

```diff
@@ -664,6 +653,82 @@ def parseFile(filename: str, belPrefix: str = "", filetype: str = "") -> Bel:
         isConfig = False
         isShared = False
 
+    if filetype == "verilog":
+        # Runs yosys on verilog file, creates netlist, saves to json in same directory.
+        json_file = filename.replace(".v", ".json")
+        runCmd = [
+            "yosys",
+            "-qp"
+            f"read_verilog {filename}; proc -noopt; write_json -compat-int {json_file}",
+        ]
+        try:
+            subprocess.run(runCmd, check=True)
+        except subprocess.CalledProcessError as e:
+            logger.error(f"Failed to run yosys command: {e}")
+            raise ValueError
+
+        with open(f"{json_file}", "r") as f:
+            data_dict = json.load(f)
+        # Default yosys list names added.
+        modules = data_dict.get("modules", {})
+        sub_modules = ["parameter_default_values", "cells", "memories", "netnames"]
+        filtered_ports: Dict[str, IO] = {}
+        # Gatheres port name and direction, filters out configbits as they show in ports.
+        for module_name, module_info in modules.items():
+            ports = module_info["ports"]
+            for port_name, details in ports.items():
+                if "ConfigBits" in port_name:
+                    continue
+                if "UserCLK" in port_name:
+                    userClk = True
+                if port_name[-1].isdigit():
+                    individually_declared = True
+                direction = IO[details["direction"].upper()]
+                bits = details.get("bits", [])
+                filtered_ports[port_name] = (direction, bits)
+        # Passed attributes dont show in port list, checks for attributes in other lists.
+        for submodule in sub_modules:
+            if submodule not in module_info:
+                continue
+            item_info = module_info.get(submodule, {})
+            for item, details in item_info.items():
+                if item == "NoConfigBits":
+                    noConfigBits = details
+                if item in filtered_ports:
+                    direction, bits = filtered_ports[item]
+                    attributes = details.get("attributes", {})
+                    for index in range(len(bits)):
+                        if len(bits) > 1:
+                            new_port_name = (
+                                f"{item}{index}"  # Multi-bit ports get index
+                            )
+                        else:
+                            new_port_name = (
+                                item  # Single-bit port uses the original name
+                            )
+                        match attributes:
```

Comment:
```
for attr in attributes:
  match attr:
    case "EXTERNAL":
      ....
```
will look better

Reply:
Trying this way causes issues, see more detail in teams.

Reply:
Change now to if else for clarity.